File: src/bugarach/detectors/coact.py
"""CoactDetect — port of interface2's ``detect_local_coincidence`` (detector #5
prototype): distinct-ROI coincidence tested against a rate-LOCAL null.

Statistic: per int_win bin, the number of DISTINCT ROIs with >= 1 event (each
ROI capped at 1, so a lone bursty ROI cannot drive it).

Null: for each candidate bin (obs >= min_rois), circular-shift each ROI's
events WITHIN a rolling context window centred on the bin, recount. This
preserves each ROI's local rate/ISI structure and randomises only cross-ROI
phase, so a window that is dense-but-unsynchronised (e.g. drug onset) gets a
null matched to the local density and is not flagged.

Significance is a one-sided Gaussian-tail p from the null's mean/sd
(p = 0.5*erfc(z/sqrt(2))), continuous so alpha can go below 1/(N+1).

Parity: the surrogate stream reproduces MATLAB exactly — MATLAB's default
``rng(seed)`` twister and numpy's ``RandomState(seed)`` generate identical
doubles, and draws are consumed in the same order (candidate bins ascending,
surrogate-major / ROI-minor within a bin). Verified against MATLAB reference
output in tests/fixtures/ref_coact_synth.json and on a real slice.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from math import erfc, sqrt

import numpy as np

from bugarach.detectors.peaks import peak_gate
from bugarach.detectors.rate import DetectorSignal


@dataclass
class CoactDetection:
    """Detected episodes + per-bin profiles (mirrors detect_local_coincidence's
    ``out``). Event fields are parallel arrays of length n_events; width_kind
    is "episode_span" (threshold mode) or "half_prominence" (peak mode)."""

    onset_sec: np.ndarray
    width_sec: np.ndarray
    strength: np.ndarray            # = z
    nrois: np.ndarray               # peak/episode-max distinct-ROI coactivity
    z: np.ndarray
    p: np.ndarray
    peak_sec: np.ndarray            # NaN in threshold mode
    t50rise: np.ndarray             # NaN in threshold mode
    t50fall: np.ndarray             # NaN in threshold mode
    width_kind: str
    ctr: np.ndarray                 # bin centres (s)
    obs: np.ndarray                 # distinct-ROI coactivity per bin
    z_prof: np.ndarray              # NaN except candidate bins
    pval_prof: np.ndarray
    nullmean_prof: np.ndarray
    signal: DetectorSignal
    ext: tuple[float, float] = (0.0, 0.0)
    opts: dict = field(default_factory=dict)

    @property
    def n_events(self) -> int:
        return self.onset_sec.size


def coact_detect(
    trains: list[np.ndarray],
    t_range: tuple[float, float],
    *,
    int_win_sec: float = 0.5,
    context_win_sec: float = 60.0,
    min_rois: int = 3,
    n_surrogates: int = 200,
    alpha: float = 0.01,
    merge_gap_sec: float = 3.0,
    rng_seed: int = 20260706,
    detection_mode: str = "threshold",
    peak_prominence: float = 0.0,
    peak_min_distance_sec: float = 0.0,
) -> CoactDetection:
    """Distinct-ROI coincidence vs a rolling rate-local circular-shift null.

    trains are per-ROI onset times (explore_sce feeds raw t50rise cells; they
    are clipped to t_range here). Defaults mirror the MATLAB function; the
    explore_sce viewer uses per-stream values (FAST: 2 s bins / 60 s context /
    alpha 1e-4; SLOW: 1 s / 120 s / 1e-6; n_surrogates 100).
    """
    if detection_mode not in ("threshold", "peak"):
        raise ValueError('detection_mode must be "threshold" or "peak"')
    rng = np.random.RandomState(rng_seed)
    t0, t1 = t_range
    bw = int_win_sec
    C = context_win_sec

    ev = []
    for v in trains:
        v = np.asarray(v, dtype=float).ravel()
        v = v[np.isfinite(v)]
        ev.append(np.sort(v[(v >= t0) & (v <= t1)]))

    nb = max(1, int(np.ceil((t1 - t0) / bw)))
    edges = t0 + np.arange(nb + 1) * bw
    ctr = t0 + (np.arange(1, nb + 1) - 0.5) * bw

    obs = np.zeros(nb)
    for e in ev:
        if e.size == 0:
            continue
        bi = np.clip(np.floor((e - t0) / bw).astype(int), 0, nb - 1)
        obs[np.unique(bi)] += 1  # distinct-ROI: 1 per ROI per bin

    cand = np.flatnonzero(obs >= min_rois)
    z = np.full(nb, np.nan)
    pval = np.full(nb, np.nan)
    nullmean = np.full(nb, np.nan)
    n_sur = int(n_surrogates)

    for b in cand:
        blo, bhi = edges[b], edges[b + 1]
        c_lo = max(t0, ctr[b] - C / 2)
        c_hi = min(t1, ctr[b] + C / 2)
        cw = c_hi - c_lo
        tlo, thi = blo - c_lo, bhi - c_lo
        ctx = []
        for e in ev:
            if e.size == 0:
                continue
            vv = e[(e >= c_lo) & (e <= c_hi)]
            if vv.size:
                ctx.append(vv - c_lo)
        # one uniform draw per (surrogate, context-ROI), surrogate-major order —
        # identical stream consumption to the MATLAB scalar-rand loops
        draws = rng.random_sample((n_sur, len(ctx)))
        counts = np.zeros(n_sur)
        for j, e in enumerate(ctx):
            shifted = np.mod(e[None, :] + draws[:, j:j + 1] * cw, cw)
            counts += ((shifted >= tlo) & (shifted < thi)).any(axis=1)
        mu = counts.mean()
        sd = counts.std(ddof=1)
        nullmean[b] = mu
        if sd > 0:
            z[b] = (obs[b] - mu) / sd
            pval[b] = 0.5 * erfc(z[b] / sqrt(2))
        elif obs[b] > mu:
            z[b], pval[b] = np.inf, 0.0
        else:
            z[b], pval[b] = 0.0, 1.0

    with np.errstate(invalid="ignore"):
        is_sig = (obs >= min_rois) & (pval <= alpha)  # NaN pval -> not significant

    if detection_mode == "peak":
        # peak-gate obs, admitting only peaks at significant bins: the
        # per-bin threshold (-Inf where significant, +Inf elsewhere) IS the
        # significance mask, and min-distance thins among significant peaks
        thrv = np.where(is_sig, -np.inf, np.inf)
        d_bins = max(1, int(np.floor(peak_min_distance_sec / bw + 0.5)))
        pk = peak_gate(obs, thrv, prominence=peak_prominence,
                       min_distance=d_bins, floor=min_rois, strict_above=True)
        onset = ctr[0] + pk.left_x * bw           # left half-prom edge, bin-centre space
        width = (pk.right_x - pk.left_x) * bw
        ev_nrois = obs[pk.idx]
        ev_z = z[pk.idx]
        ev_p = pval[pk.idx]
        peak_sec = ctr[0] + pk.idx * bw
        t50rise = onset.copy()
        t50fall = onset + width
        width_kind = "half_prominence"
    else:
        sig = cand[is_sig[cand]]
        starts_b, ends_b = [], []
        if sig.size:
            s_cur = prev = sig[0]
            for b in sig[1:]:
                if edges[b] - edges[prev + 1] <= merge_gap_sec:
                    prev = b
                else:
                    starts_b.append(s_cur)
                    ends_b.append(prev)
                    s_cur = prev = b
            starts_b.append(s_cur)
            ends_b.append(prev)
        n = len(starts_b)
        onset = np.array([edges[s] for s in starts_b])
        width = np.array([edges[e + 1] - edges[s] for s, e in zip(starts_b, ends_b)])
        ev_nrois = np.array([obs[s:e + 1].max() for s, e in zip(starts_b, ends_b)])
        # merged episodes span non-candidate bins whose z/p are NaN; MATLAB's
        # max/min ignore NaN, so aggregate with the nan-variants
        ev_z = np.array([np.nanmax(z[s:e + 1]) for s, e in zip(starts_b, ends_b)])
        ev_p = np.array([np.nanmin(pval[s:e + 1]) for s, e in zip(starts_b, ends_b)])
        peak_sec = np.full(n, np.nan)
        t50rise = np.full(n, np.nan)
        t50fall = np.full(n, np.nan)
        width_kind = "episode_span"

    opts = {
        "int_win_sec": int_win_sec, "context_win_sec": context_win_sec,
        "min_rois": min_rois, "n_surrogates": n_surrogates, "alpha": alpha,
        "merge_gap_sec": merge_gap_sec, "rng_seed": rng_seed,
        "detection_mode": detection_mode, "peak_prominence": peak_prominence,
        "peak_min_distance_sec": peak_min_distance_sec,
    }
    return CoactDetection(
        onset_sec=onset, width_sec=width, strength=ev_z.copy(), nrois=ev_nrois,
        z=ev_z, p=ev_p, peak_sec=peak_sec, t50rise=t50rise, t50fall=t50fall,
        width_kind=width_kind,
        ctr=ctr, obs=obs, z_prof=z, pval_prof=pval, nullmean_prof=nullmean,
        signal=DetectorSignal(t=ctr, y=obs, ref=nullmean, threshold=None,
                              hilite=np.empty((0, 2)),
                              name="coact / local null", kind="local_coincidence"),
        ext=t_range, opts=opts,
    )

File: docs/site/detectors/coact.js
/* coact — CoactDetect, as an object.

   ONE FILE IS ONE DETECTOR: the descriptor the page renders from, and the code
   that runs. Delete this file and coact is gone — picker, controls and algorithm
   together — which is the point of the folder (ADR-0005).

   Assembled into the page at BUILD time, not load time: docs/site is one
   self-contained file making zero requests, so it works from file:// and no
   recording leaves the machine. build_site.py's NETWORK guard enforces it.

   WHAT THE SHELL PROVIDES and this file uses: the shared helpers. `clipSorted`
   is called by five detectors and `matlabPrctile` by three, so they stay in the
   shell — a shared helper living in one detector's file would make deleting
   that file break the others.

   PARITY: `bugarach.detectors.coact` is the other implementation of this, and
   tests/test_webapp_coact_detect_parity.py pins the two. A change here that is
   not made there is a divergence, not an improvement.
*/

// ---- descriptor -------------------------------------------------------------
registerDetector(
{
    key: "coact",
    label: "CoactDetect",
    blurb: "how unlikely",
    ctl: "dCoactCtl", fine: "dFineCoact",
    read: dt => ({ gridDt: dt,
      intWinSec: clamp(numById("cWin"), 0.1, 60),
      contextWinSec: clamp(numById("cCtx"), 10, 1200),
      alpha: clamp(numById("cAlpha"), 1e-7, 0.5),
      nSurrogates: Math.round(clamp(numById("cSur"), 20, 1000)) }),
    run: (trains, range, cfg) => coactDetect(trains, range, cfg),
    settings: cfg => "bin " + cfg.intWinSec + " s · context " + cfg.contextWinSec
      + " s · alpha " + cfg.alpha + " per bin · " + cfg.nSurrogates
      + " shuffles · floor 3 ROIs",
    params: { intWinSec: "cWin", contextWinSec: "cCtx", alpha: "cAlpha",
              nSurrogates: "cSur" },
    knob: { key: "alpha", input: "cAlpha", name: "alpha", unit: "per bin",
            scale: "log", grid: [1e-2, 1e-3, 1e-4, 1e-5, 1e-6, 1e-7] },
  }
);

// ---- algorithm --------------------------------------------------------------
function coactDetect(trains, range, opts) {
  const o = opts || {};
  const bw = o.intWinSec === undefined ? 0.5 : o.intWinSec;
  const C = o.contextWinSec === undefined ? 60.0 : o.contextWinSec;
  const minRois = o.minRois === undefined ? 3 : o.minRois;
  const nSur = o.nSurrogates === undefined ? 200 : o.nSurrogates;
  const alpha = o.alpha === undefined ? 0.01 : o.alpha;
  const mgap = o.mergeGapSec === undefined ? 3.0 : o.mergeGapSec;
  const u = rng32(o.seed === undefined ? 20260706 : o.seed);

  const t0 = range[0], t1 = range[1];
  const ev = clipSorted(trains, t0, t1);
  const nb = Math.max(1, Math.ceil((t1 - t0) / bw));
  const edges = new Float64Array(nb + 1);
  for (let k = 0; k <= nb; k++) edges[k] = t0 + k * bw;
  const ctr = new Float64Array(nb);
  for (let k = 0; k < nb; k++) ctr[k] = t0 + (k + 0.5) * bw;

  // distinct ROIs per bin — one per ROI per bin, however often it fired
  const obs = new Float64Array(nb);
  for (const e of ev) {
    if (!e.length) continue;
    const seen = new Set();
    for (const x of e) {
      let b = Math.floor((x - t0) / bw);
      if (b < 0) b = 0; else if (b > nb - 1) b = nb - 1;
      seen.add(b);
    }
    for (const b of seen) obs[b] += 1;
  }

  const z = new Float64Array(nb).fill(NaN);
  const pval = new Float64Array(nb).fill(NaN);
  const nullmean = new Float64Array(nb).fill(NaN);
  const cand = [];
  for (let b = 0; b < nb; b++) if (obs[b] >= minRois) cand.push(b);

  if (o.pvals) {
    for (let b = 0; b < nb; b++) pval[b] = o.pvals[b];
  } else {
    for (const b of cand) {
      const blo = edges[b], bhi = edges[b + 1];
      const cLo = Math.max(t0, ctr[b] - C / 2);
      const cHi = Math.min(t1, ctr[b] + C / 2);
      const cw = cHi - cLo;
      const tlo = blo - cLo, thi = bhi - cLo;
      const ctx = [];
      for (const e of ev) {
        if (!e.length) continue;
        const vv = [];
        for (const x of e) if (x >= cLo && x <= cHi) vv.push(x - cLo);
        if (vv.length) ctx.push(vv);
      }
      // one draw per (surrogate, context ROI), surrogate-major — the same stream
      // shape the MATLAB scalar-rand loops consume
      const counts = new Float64Array(nSur);
      const draws = new Float64Array(nSur * ctx.length);
      for (let i = 0; i < draws.length; i++) draws[i] = u();
      for (let j = 0; j < ctx.length; j++) {
        for (let s = 0; s < nSur; s++) {
          const d = draws[s * ctx.length + j];
          let hit = false;
          for (const x of ctx[j]) {
            let sh = (x + d * cw) % cw;
            if (sh < 0) sh += cw;
            if (sh >= tlo && sh < thi) { hit = true; break; }
          }
          if (hit) counts[s] += 1;
        }
      }
      const st = coactStats(obs[b], counts);
      nullmean[b] = st.mu; z[b] = st.z; pval[b] = st.p;
    }
  }

  // a NaN p-value is not significant — the bin was never a candidate
  const sig = [];
  for (const b of cand) if (pval[b] <= alpha) sig.push(b);

  const startsB = [], endsB = [];
  if (sig.length) {
    let sCur = sig[0], prev = sig[0];
    for (let k = 1; k < sig.length; k++) {
      const b = sig[k];
      // gap measured from the END of the previous significant bin, not its start
      if (edges[b] - edges[prev + 1] <= mgap) prev = b;
      else { startsB.push(sCur); endsB.push(prev); sCur = prev = b; }
    }
    startsB.push(sCur); endsB.push(prev);
  }

  const starts = [], widths = [], ends = [], nrois = [], evZ = [], evP = [];
  for (let k = 0; k < startsB.length; k++) {
    const s = startsB[k], e = endsB[k];
    starts.push(edges[s]);
    widths.push(edges[e + 1] - edges[s]);
    ends.push(edges[e + 1]);
    let mx = -Infinity, zmax = NaN, pmin = NaN;
    for (let b = s; b <= e; b++) {
      if (obs[b] > mx) mx = obs[b];
      // merged episodes span non-candidate bins whose z/p are NaN, and MATLAB's
      // max/min ignore NaN — so these are the nan-variants, not the plain ones
      if (!Number.isNaN(z[b])) zmax = Number.isNaN(zmax) ? z[b] : Math.max(zmax, z[b]);
      if (!Number.isNaN(pval[b])) pmin = Number.isNaN(pmin) ? pval[b] : Math.min(pmin, pval[b]);
    }
    nrois.push(mx); evZ.push(zmax); evP.push(pmin);
  }
  return { starts, ends, widths, nrois, z: evZ, p: evP,
           ctr, obs, zProf: z, pvalProf: pval, nullmean,
           nEvents: starts.length };
}
